File: backend/schemas/granja/Instalaciones.py
// src/services/granjaInstalacionesService.ts
export type InstalacionGranja = {
  id: string;
  nombre: string;
  tipo: "Galpón" | "Depósito" | "Oficina" | "Corral" | "Enfermería" | "Otro";
  superficieM2: number;
  capacidad: string;
  estado: "Operativa" | "Mantenimiento" | "Inactiva";
  descripcion: string;
  ubicacionZona: string; // Puede referenciar id/nombre de zona de Infraestructura
  observaciones: string;
};

const STORAGE_KEY = "granja_instalaciones";

function getAll(): InstalacionGranja[] {
  const raw = localStorage.getItem(STORAGE_KEY);
  return raw ? JSON.parse(raw) : [];
}

function saveAll(lista: InstalacionGranja[]) {
  localStorage.setItem(STORAGE_KEY, JSON.stringify(lista));
}

export function getInstalaciones() {
  return getAll();
}

export function addInstalacion(inst: Omit<InstalacionGranja, "id">) {
  const nueva: InstalacionGranja = { ...inst, id: crypto.randomUUID() };
  const instalaciones = getAll();
  instalaciones.push(nueva);
  saveAll(instalaciones);
  return nueva;
}

export function updateInstalacion(id: string, inst: Omit<InstalacionGranja, "id">) {
  const instalaciones = getAll();
  const idx = instalaciones.findIndex(i => i.id === id);
  if (idx >= 0) {
    instalaciones[idx] = { ...inst, id };
    saveAll(instalaciones);
    return instalaciones[idx];
  }
  throw new Error("Registro no encontrado");
}

export function deleteInstalacion(id: string) {
  const instalaciones = getAll().filter(i => i.id !== id);
  saveAll(instalaciones);
}

File: frontend/src/services/granja/Instalaciones.ts
import API_BASE from "../../config/api";

export type InstalacionGranja = {
  id: number;
  nombre: string;
  tipo: "Galpón" | "Depósito" | "Oficina" | "Corral" | "Enfermería" | "Otro";
  superficieM2: number;
  capacidad: string;
  estado: "Operativa" | "Mantenimiento" | "Inactiva";
  descripcion: string;
  ubicacionZona: string;
  observaciones: string;
};

export type NuevaInstalacionGranja = Omit<InstalacionGranja, "id">;

const EMPRESA_ID = 1;
const GRANJA_ID = 1;

function mapApiToInst(api: any): InstalacionGranja {
  return {
    id: api.id,
    nombre: api.nombre,
    tipo: api.tipo,
    superficieM2: api.superficieM2,
    capacidad: api.capacidad,
    estado: api.estado,
    descripcion: api.descripcion,
    ubicacionZona: api.ubicacionZona,
    observaciones: api.observaciones ?? "",
  };
}

export async function getInstalaciones(): Promise<InstalacionGranja[]> {
  const res = await fetch(
    `${API_BASE}/granja/instalaciones/?empresa_id=${EMPRESA_ID}&granja_id=${GRANJA_ID}`
  );
  if (!res.ok) {
    throw new Error("Error al obtener instalaciones de la granja");
  }
  const data = await res.json();
  return data.map(mapApiToInst);
}

export async function addInstalacion(
  inst: NuevaInstalacionGranja
): Promise<InstalacionGranja> {
  const payload = {
    empresa_id: EMPRESA_ID,
    granja_id: GRANJA_ID,
    nombre: inst.nombre,
    tipo: inst.tipo,
    superficieM2: inst.superficieM2,
    capacidad: inst.capacidad,
    estado: inst.estado,
    descripcion: inst.descripcion,
    ubicacionZona: inst.ubicacionZona,
    observaciones: inst.observaciones,
  };

  const res = await fetch(`${API_BASE}/granja/instalaciones/`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!res.ok) {
    throw new Error("Error al registrar instalación");
  }
  const data = await res.json();
  return mapApiToInst(data);
}

export async function updateInstalacion(
  id: number,
  inst: NuevaInstalacionGranja
): Promise<InstalacionGranja> {
  const payload = {
    nombre: inst.nombre,
    tipo: inst.tipo,
    superficieM2: inst.superficieM2,
    capacidad: inst.capacidad,
    estado: inst.estado,
    descripcion: inst.descripcion,
    ubicacionZona: inst.ubicacionZona,
    observaciones: inst.observaciones,
  };

  const res = await fetch(
    `${API_BASE}/granja/instalaciones/${id}?empresa_id=${EMPRESA_ID}&granja_id=${GRANJA_ID}`,
    {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    }
  );
  if (!res.ok) {
    throw new Error("Error al actualizar instalación");
  }
  const data = await res.json();
  return mapApiToInst(data);
}

export async function deleteInstalacion(id: number): Promise<void> {
  const res = await fetch(
    `${API_BASE}/granja/instalaciones/${id}?empresa_id=${EMPRESA_ID}&granja_id=${GRANJA_ID}`,
    { method: "DELETE" }
  );
  if (!res.ok) {
    throw new Error("Error al eliminar instalación");
  }
}
